src/Sylius/Component/Core/Taxation/OrderItemsByZoneTaxesApplicator.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Taxation;
+
+use Sylius\Bundle\CoreBundle\Distributor\IntegerDistributorInterface;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Order\Factory\AdjustmentFactoryInterface;
+use Sylius\Component\Taxation\Calculator\CalculatorInterface;
+use Sylius\Component\Taxation\Resolver\TaxRateResolverInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class OrderItemsByZoneTaxesApplicator implements OrderTaxesByZoneApplicatorInterface
+{
+    /**
+     * @var CalculatorInterface
+     */
+    private $calculator;
+
+    /**
+     * @var AdjustmentFactoryInterface
+     */
+    private $adjustmentFactory;
+
+    /**
+     * @var IntegerDistributorInterface
+     */
+    private $distributor;
+
+    /**
+     * @var TaxRateResolverInterface
+     */
+    private $taxRateResolver;
+
+    /**
+     * @param CalculatorInterface $calculator
+     * @param AdjustmentFactoryInterface $adjustmentFactory
+     * @param IntegerDistributorInterface $distributor
+     * @param TaxRateResolverInterface $taxRateResolver
+     */
+    public function __construct(CalculatorInterface $calculator, AdjustmentFactoryInterface $adjustmentFactory, IntegerDistributorInterface $distributor, TaxRateResolverInterface $taxRateResolver)
+    {
+        $this->calculator = $calculator;
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->distributor = $distributor;
+        $this->taxRateResolver = $taxRateResolver;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order, ZoneInterface $zone)
+    {
+        foreach ($order->getItems() as $item) {
+            $taxRate = $this->taxRateResolver->resolve($item->getProduct(), array('zone' => $zone));
+
+            if (null === $taxRate) {
+                continue;
+            }
+
+            $units = $item->getUnits();
+            if ($units->isEmpty()) {
```

Comment:
OK for now but we need to rethink the behavior of ``isEmpty`` method on the order. Normally we should use it here, but it checks if there are any items.